idaes/core/util/model_diagnostics.py

```diff
@@ -0,0 +1,616 @@
+                else:
+                    # Pretty print constraint
+                    print("\ncount =",i,"\t|residual| =",r)
+                    c.pprint()
+        
+                if print_level == 2:
+                    # print values and bounds for each variable in the constraint
+                    print("variable\tlower\tvalue\tupper")
+                    for v in identify_variables(c.body):
+                        self.print_variable_bounds(v)
+            else:
+                print("No constraints with residuals larger than",tol,"!")
+                
+        return lrs
+
+    def check_variable_bounds(self,tol=1e-5, relative=False, skip_lb = False, skip_ub = False, verbose=True):
+        """
+        Return a ComponentSet of all variables within a tolerance of their bounds.
+        Args:
+            block : model to be studied
+            tol : residual threshold for inclusion in ComponentSet (default = 1e-5)
+            relative : Boolean, use relative tolerance (default = False)
+            skip_lb: Boolean to skip lower bound (default = False)
+            skip_ub: Boolean to skip upper bound (default = False)
+            verbose: Boolean to toggle on printing to screen (default = True)
+        Returns:
+            A ComponentSet including all Constraint components with a residual
+            greater than tol which appear in block
+        """
+        vnbs = variables_near_bounds_set(self.block, tol, relative, skip_lb, skip_ub)
+        
+        if verbose:
+            print(" ")
+            if relative:
+                s = "(relative)"
+            else:
+                s = "(absolute)"
+            if len(vnbs) > 0:
+
+                print("Variables within",tol,s,"of their bounds:")
+                print("variable\tlower\tvalue\tupper")
+                for v in vnbs:
+                    self.print_variable_bounds(v)
+            else:
+                print("No variables within",tol,s,"of their bounds.")
+            
+        return vnbs
+    
+    def check_rank_equality_constraints(self, tol=1E-6):
+        """
+        Method to check the rank of the Jacobian of the equality constraints
+        
+        Args:
+            tol: Tolerance for smallest singular value (default=1E-6)
+        
+        Returns:
+            Number of singular values less than tolerance (-1 means error)
+        
+        """
+        
+        print("\nChecking rank of Jacobian of equality constraints...")
+        
+        print("Model contains",self.n_eq,"equality constraints and",
+                    self.n_var,"variables.")
+        
+        counter = 0
+        if self.n_eq > 1:
+            if self.s is None:
+                self.svd_analysis()
+
+            n = len(self.s)
+        
+            print("Smallest singular value(s):")
+            for i in range(n):
+                print("%.3E" % self.s[i])
+                if self.s[i] < tol:
+                    counter += 1
```

Comment:
Out of curiosity, is there any reason you check the singular values themselves, and not the ratios of these singular values to the max singular value?

Reply:
I have not given it too much thought. Perhaps using absolute singular values tells you more about the current model formulations. I wonder if scaling by the max singular value would hide scaling issues. The counterpoint is that the condition number is the ratio of the largest and smallest singular values.